### PR TITLE
Rework frame pipelining and the animator to make it easier to wait of fences for frame completion.

### PR DESCRIPTION
**Pull request by chinmaygarde**
cc @abarth @jason-simmons 

I am not going to land this just yet. Have a another patch that deals with platform specific fences. But the basic concept is the same.

Have also cleaned up the animator a bit. Will keep stress testing it some more but it seems to be holding up pretty well.

Also, `VSync` providers are now mandatory. But I realize not all platform may/can have actual vsync providers. So a fallback provider is not included. The mojo frame scheduler needs to be moved behind the same. Will do that in a final patch.

### CODE REVIEW

File: sky/shell/gpu/direct/rasterizer_direct.cc [hunk 1]
```diff
@@ -86,33 +81,36 @@ flow::LayerTree* RasterizerDirect::GetLastLayerTree() {
...
     return;
   }
 
-  std::unique_ptr<flow::LayerTree> layer_tree(
-      reinterpret_cast<flow::LayerTree*>(layer_tree_ptr));
+  flow::Pipeline<flow::LayerTree>::Consumer consumer =
+      std::bind(&RasterizerDirect::DoDraw, this, std::placeholders::_1);
+
+  // Consume as many pipeline items as possible.
+  while (pipeline->consume(consumer, ftl::TimeDelta::Zero()))
```
Comment (chinmaygarde) [inline sky/shell/gpu/direct/rasterizer_direct.cc]: ugh, this is incorrect. Updating.

File: flow/latch.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  // already zero, this operation is a no-op
+  void wait();
+
+  // Decrement the count of the latch
+  void countDown();
+
+ private:
+  std::atomic_size_t count_;
+  std::condition_variable condition_;
+  std::mutex condition_mutex_;
```
Comment (abarth) [inline flow/latch.h]: Is there a reason to use the std versions instead of the ftl versions?

File: flow/pipeline.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+#include <queue>
+#include <type_traits>
+
+namespace flow {
+
+template <class R>
+class Pipeline {
+ public:
+  using Resource = R;
+  using ResourceRef = std::unique_ptr<Resource>;
```
Comment (abarth) [inline flow/pipeline.h]: ResourcePtr?

File: flow/pipeline.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+  Pipeline(uint32_t depth) : empty_(depth), available_(0) {}
```
Comment (abarth) [inline flow/pipeline.h]: explicit

File: flow/semaphore.cc [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    if (_sem == nullptr) {
+      return false;
+    }
+
+    auto dispatch_timeout =
+        dispatch_time(DISPATCH_TIME_NOW, timeout == ftl::TimeDelta::Max()
+                                             ? DISPATCH_TIME_FOREVER
+                                             : timeout.ToNanoseconds());
+
+    return dispatch_semaphore_wait(_sem, dispatch_timeout) == 0;
```
Comment (abarth) [inline flow/semaphore.cc]: Are you sure these timebases are the same?

File: flow/semaphore.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+#include "lib/ftl/time/time_delta.h"
+#include "lib/ftl/compiler_specific.h"
+
+namespace flow {
+
+class PlatformSemaphore;
+
+class Semaphore {
+ public:
+  Semaphore(uint32_t count);
```
Comment (abarth) [inline flow/semaphore.h]: explicit

File: services/vsync/fallback/vsync_provider_fallback_impl.cc [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  if (armed_) {
+    return;
+  }
+
+  armed_ = true;
+
+  const base::TimeDelta interval = base::TimeDelta::FromSecondsD(1.0 / 60.0);
+
+  const base::TimeTicks now = base::TimeTicks::Now();
+  const base::TimeTicks next = now.SnappedToNextTick(phase_, interval);
```
Comment (abarth) [inline services/vsync/fallback/vsync_provider_fallback_impl.cc]: I'm not sure how important quality is on platforms that lack a vsync signal, but this isn't the highest quality algorithm.  It's probably fine for our purposes though.

**Comment by abarth** [comment]
I'm not sure I understand the general approach.  Is the idea that the GPU thread will sit in the semaphore rather than in the event loop?  Also, is the idea to block the UI thread on the semaphore if it gets too far ahead?  I'm concerned about both those things, if true.

File: flow/latch.h (hunk 2 above)
Comment (chinmaygarde) [inline flow/latch.h]: No. Will update.

File: flow/pipeline.h (hunk 4 above)
Comment (chinmaygarde) [inline flow/pipeline.h]: Done.

File: flow/pipeline.h (hunk 3 above)
Comment (chinmaygarde) [inline flow/pipeline.h]: Done.

File: flow/semaphore.cc (hunk 5 above)
Comment (chinmaygarde) [inline flow/semaphore.cc]: * `dispatch_time` works with nanoseconds.

File: flow/semaphore.h (hunk 6 above)
Comment (chinmaygarde) [inline flow/semaphore.h]: Done.

File: services/vsync/fallback/vsync_provider_fallback_impl.cc (hunk 7 above)
Comment (chinmaygarde) [inline services/vsync/fallback/vsync_provider_fallback_impl.cc]: iOS, Android & Mac all have actual VSync providers. So the only platform that might have to use the fallback provider is Linux Desktop.

Though I agree this is a suboptimal solution, it does not accumulate drift. I believe the old solution would schedule the next `BeginFrame` a full frame interval after the `RequestFrame` (and not account for any portion of the time slice already consumed before the `RequestFrame`). What would be a higher quality algorithm without an actual VSync provider?

**Comment by chinmaygarde** [comment]
> Also, is the idea to block the UI thread on the semaphore if it gets too far ahead?

No, the call to `flow::Pipeline::produce` is made with a zero timeout. If the pipeline cannot immediately service the production of the layer tree, the caller will try again at the next VSync.

> Is the idea that the GPU thread will sit in the semaphore rather than in the event loop?

No, The call to `flow::Pipeline::consume` is also made with a zero timeout. What is currently incorrect (something I mentioned in my comment in the morning) it the desire to consume all pipeline items as soon as possible is accomplished in a tight loop. I need to yield to the platform event loop before consuming the next item as soon as possible.

We cannot ever sit in anything other than the platform event loop. This would cause all post task invocations to back up till the next time the rasterizer idles (which may never happen if the work items are expensive enough).

File: flow/semaphore.cc (hunk 5 above)
Comment (abarth) [inline flow/semaphore.cc]: Ah, delta.  Yes, you're right.

**Comment by abarth** [comment]
Excellent.  Thanks!

**Comment by chinmaygarde** [comment]
Updated to fix the issue described in 
screen shot 2016-08-16 at 2 57 25 pm

Also accounted for the case in which callers (both Dart and Native) may call the RequestFrame multiple times in a tight loop.

**Comment by chinmaygarde** [comment]
This patch is done. Closing. Will send smaller patches separately.

**Pull request closed without merging**